design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+# KLIP 8 - Interactive Queries
```

Comment:
Can we pleeeeeease come up with a better name that IQ...? There is nothing more _interactive_ about these queries that other queries.

I know this isn't crucial at this point, but the sooner we come up with a better name, i.e. the one we want to use externally, the more consistently we'll use it internally in code/docs/discussions. Also, this KLIP is public, so would be good to use the name this will be released as.

So let's discuss nomenclature...

We currently use the terminology:

*  _transient query_, which is a bare `SELECT * FROM FOO;` where `FOO` is a stream or table, i.e. a query that isn't  persistent over a machine restart.  But without a limit clause the query will run continuously.
* _persistent query_, which is a `CREATE TABLE/STREAM AS` or `INSERT INTO` statement, which are persistent over a machine restart and run continuously.

I'd like to propose we also consider introducing:

* _continuous query_, which is a query that runs without end. Both the current _transient_ and _persistent_ queries are _continuous_, unless they have a limit clause.
* _terminal query_, which is the opposite of _continuous_.  This would include any _streaming_ query with a limit clause and these new _interactive_ queries. 

And:

* _streaming query_, which is what we currently support, i.e. stream me back the stream of events or table updates.

Which then brings us to what we should call IQ.  What's the opposite of a _streaming_ query?  Maybe:

* _non-streaming query_
* _normal query_ / _standard query_
* _table query_ (none-streaming only makes sense on a table), though this may become muddy when talking about streaming table queries vs table queries.
* _tranditional query_
* _ conventional query_

I'm not particularly fixed on any of these, but I think either _non-streaming_ or _standard_ queries is better than _interactive_.

Thoughts?

Reply:
I'm still working through all of the context around this and related work, so take what I say with a grain of salt :). IIUC interactive queries correctly, I've been using the placeholder "ad hoc queries" in my mind.

Reply:
@big-andy-coates definitely open to a better name than IQ. I personally don't love it either. Point-in-time queries seems pretty reasonable, or just point queries. Terminal query is also interesting...

Reply:
@cpettitt-confluent that is definitely a reasonable take! I think the reason we haven't gone with "adhoc queries" is that KSQL already supports a kind of adhoc query. We call these "transient" queries, which are just streaming queries that aren't persisted. For example,

```
ksql> SELECT * FROM my_stream;
+------+-------+
|   x  |   y   |
+------+-------+
|   0  |   1   |
|   1  |   2   |
...
(new rows are output as they are written to my_stream)
Ctrl-C
Query terminated
```
As soon as the user terminates the transient query with Ctrl-C it no longer exists. But perhaps it is worth considering using the term adhoc for IQ. It's a very well-understood name that does represent what IQ does pretty well.

Reply:
I was starting to get comfortable with "continuous" and "point-in-time", for what it's worth. "transient" is still a little left field, but it works alright.

Reply:
I'm liking point-in-time queries or point queries for short.

Reply:
+1 to "continuous" and "point-in-time" queries.